Speculative mitigation against a crash involving CodeBlock::m_metadata.

MenloDorian opened the pull request:
#### 59012130505d6145ec4e7a8dc6e718106a8baba7
<pre>
Speculative mitigation against a crash involving CodeBlock::m_metadata.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274691">https://bugs.webkit.org/show_bug.cgi?id=274691</a>
<a href="https://rdar.apple.com/127571559">rdar://127571559</a>

Reviewed by Yijia Huang and Yusuke Suzuki.

We sometimes see crashes due to CodeBlock::m_metadata-&gt;unlinkedMetadata() being null.
This change adds a speculative null check to mitigate against stepping on this.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::~CodeBlock):
* Source/JavaScriptCore/bytecode/MetadataTable.cpp:
(JSC::MetadataTable::destroy):
* Source/JavaScriptCore/bytecode/MetadataTable.h:
(JSC::MetadataTable::isDestroyed const):

Canonical link: <a href="https://commits.webkit.org/279300@main">https://commits.webkit.org/279300@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75399ba9223a48f01bedeece35c85bffc5361c49

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53083 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5570 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/56362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/3806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3533 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/56362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/3806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55181 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/30178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/56362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3145 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/1965 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46439 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3303 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57957 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52596 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/28224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46043 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64901 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7799 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29198 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12317 "Found 10 new JSC stress test failures: stress/proxy-set-prototype-of.js.bytecode-cache, stress/setter-frame-flush.js.default, stress/spread-non-array.js.dfg-eager-no-cjit-validate, stress/spread-non-array.js.mini-mode, wasm.yaml/wasm/function-tests/add-12.js.wasm-eager, wasm.yaml/wasm/function-tests/i32-load8-s.js.wasm-eager, wasm.yaml/wasm/js-api/global-mutate.js.wasm-eager, wasm.yaml/wasm/stress/immutable-globals.js.wasm-eager, wasm.yaml/wasm/stress/js-to-wasm-many-double-from-js.js.wasm-eager, wasm.yaml/wasm/stress/too-many-locals.js.wasm-eager (failure)") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me

hyjorc1 commented on Source/JavaScriptCore/bytecode/MetadataTable.cpp:
```diff
@@ -61,16 +61,17 @@ MetadataTable::~MetadataTable()
 
 void MetadataTable::destroy(MetadataTable* table)
```
Do we need to assert the table pointer even it seems to be only used in `deref` function?

MenloDorian replied:
We did not see a crash scenario where the table pointer was null.  So, I don't want to add extra mitigation for something that we have no evidence of.

Constellation commented on Source/JavaScriptCore/bytecode/MetadataTable.cpp:
```diff
@@ -61,16 +61,17 @@ MetadataTable::~MetadataTable()
     }
 
+    RefPtr<UnlinkedMetadataTable> unlinkedMetadata = WTFMove(table->linkingData().unlinkedMetadata);
+    ASSERT(table->isDestroyed());
```
Is it correct?

hyjorc1 replied:
I think so, move the `unlinkedMetadata` should just replace that with a nullptr then it is `isDestroyed`.

MenloDorian replied:
Yes, the WTFMove basically exchanges the Ref with a nullptr.  This invokes Ref::leakRef() which nullifies the Ref before returning the leaked ptr.

hyjorc1 commented on Source/JavaScriptCore/bytecode/MetadataTable.cpp:
```diff
@@ -61,16 +61,17 @@ MetadataTable::~MetadataTable()
         return;
     }
 
+    RefPtr<UnlinkedMetadataTable> unlinkedMetadata = WTFMove(table->linkingData().unlinkedMetadata);
```
So, there is a case that the `MetadataTable` was valid however its `unlinkedMetadata` was gone. That's why we need to check `isDestroyed` first here.

MenloDorian replied:
`unlinkedMetadata` is exchanged with null here, and this should be the only place that it is set to null.  The pre-existing code was doing the null check after the swap out.  It was also calling `table->~MetadataTable()` regardless.  This doesn't make sense.  Destruction should only happen once.  If there's some way that we can get to this function more than once, then the 2nd time should not execute any of this code.  Hence, I moved the null check to the top.

hyjorc1 commented on Source/JavaScriptCore/bytecode/CodeBlock.cpp:
```diff
@@ -852,12 +852,11 @@ CodeBlock::~CodeBlock()
-        if (m_metadata) {
+        // FIXME: This check should really not be necessary, see https://webkit.org/b/272787
+        ASSERT(!m_metadata || m_metadata->unlinkedMetadata());
+        if (m_metadata && !m_metadata->isDestroyed()) {
```
Same here in order to mitigate the unexpected case that the `MetadataTable` was valid but the `unlinkedMetadata` was gone.

MenloDorian replied:
We saw the case where `m_metadata` was not null, but the `unlinkedMetadata` was null.  It is unexplained as to how this happens.  Hence, this is a speculative fix.

hyjorc1 reviewed (approved):
r=me, too.

Commit:
Speculative mitigation against a crash involving CodeBlock::m_metadata.
https://bugs.webkit.org/show_bug.cgi?id=274691
rdar://127571559

Reviewed by Yijia Huang and Yusuke Suzuki.

We sometimes see crashes due to CodeBlock::m_metadata->unlinkedMetadata() being null.
This change adds a speculative null check to mitigate against stepping on this.

* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::~CodeBlock):
* Source/JavaScriptCore/bytecode/MetadataTable.cpp:
(JSC::MetadataTable::destroy):
* Source/JavaScriptCore/bytecode/MetadataTable.h:
(JSC::MetadataTable::isDestroyed const):

Canonical link: https://commits.webkit.org/279300@main
